File: services/app/dockers/ts/checker_example.ts
import solution from './solution_example';
import { assert } from 'chai';

let success: boolean = true;

try {
  const assertEqual = function (result, expected, errorMessage) {
    try {
      assert.deepEqual(result, expected);

      process.stdout.write(`${JSON.stringify({
        status: 'success',
        result,
      })}\n`);
    } catch (err) {
      process.stdout.write(`${JSON.stringify({
        status: 'failure',
        result,
        arguments: errorMessage,
      })}\n`);
      success = false;
    }
  };

  const a1: number = 1;
  const b1: number = 2;
  const expected1: number = 3;

  assertEqual(solution(a1, b1), expected1, [1, 2]);

  const a2 = 5;
  const b2 = 3;
  const expected2: number = 8;

  assertEqual(solution(a2, b2), expected2, [5, 3]);

  if (success) {
    process.stdout.write(`${JSON.stringify({
      status: 'ok',
      result: '__code-0__',
    })}\n`);
  }
  process.exit(0);

} catch (err) {
  process.stdout.write(`${JSON.stringify({
    status: 'failure',
    result: err.message,
  })}\n`);
  process.exit(0);
}

File: services/app/dockers/golang/checker_example.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"reflect"
)

func main() {
	defer func() {
		if err := recover(); err != nil {
			sendMessage("error", fmt.Sprintf("%s", err), nil)
		}
		os.Exit(0)
	}()

	success := true

	var a1 int64 = 1
	var b1 int64 = 2
	var expected1 int64 = 3

	success = assertSolution(solution(a1, b1), expected1, []interface{}{a1, b1}, success)

	var a2 int64 = 3
	var b2 int64 = 2
	var expected2 int64 = 5

	success = assertSolution(solution(a2, b2), expected2, []interface{}{a2, b2}, success)

	if success {
		sendMessage("ok", "__code-0__", nil)
	}
}

type resultMessage struct {
	Status    string      `json:"status"`
	Result    interface{} `json:"result"`
	Arguments interface{} `json:"arguments,omitempty"`
}

func assertSolution(result, expected, message interface{}, success bool) bool {
	status := reflect.DeepEqual(result, expected)
	if !status {
		sendMessage("failure", result, message)
		return false
	}

	return success
}

func sendMessage(status string, result, arguments interface{}) {
	message := resultMessage{
		Status:    status,
		Result:    result,
		Arguments: arguments,
	}

	jsonMessage, err := json.Marshal(message)
	if err != nil {
		fmt.Println("Marshaler error")
		os.Exit(0)
	}

	fmt.Print(string(jsonMessage))
}
